backfill/2025-01-22-firefox-installer schema-error/README.md

```diff
@@ -0,0 +1,186 @@
+valid errors.  These errors account for 4.7% of the reingested stub installer pings.
+
+The new fields are correctly being populated, matching the values in the payload (for full installer) or the URI (for stub):
+```sql
+SELECT
+  COUNT(*),
+  ping_version,
+  windows_ubr,
+  stub_build_id,
+FROM
+  `moz-fx-data-backfill-1.firefox_installer_live.install_v1`
+WHERE
+  DATE(submission_timestamp) < "2025-01-23"
+GROUP BY ALL
+```
+
+Using [null_field_ratio.sql](./null_field_ratio.sql), the null ratios approximately match the live data.
+
+## Dedupe
+
+We need to dedupe by document id within `moz-fx-data-backfill-1.firefox_installer_live.install_v1` and also 
+make sure the document ids aren't already in `moz-fx-data-shared-prod.firefox_installer_stable.install_v1`.
+We will maintain the same semantics, deduping doc ids per day, not across the whole backfill interval.  
+
+We can see that there are 2026 duplicates within the backfill data ([duplicates1.sql](./duplicates1.sql))
+and there are an additional 95 rows that are duplicated in the stable table ([duplicates2.sql](./duplicates2.sql)).
+A spot check shows that the one in the stable table are actually different pings with different data, but they're considered duplicates anyway.
+
+Backfill query (ensure `moz-fx-data-backfill-1.firefox_installer_stable.install_v1` is empty first):
+```sql
+WITH backfill AS (
+  SELECT
+    *
+  FROM
+    `moz-fx-data-backfill-1.firefox_installer_live.install_v1`
+  WHERE
+    DATE(submission_timestamp) BETWEEN "2024-12-30" AND "2025-01-22"
+  QUALIFY
+    ROW_NUMBER() OVER (PARTITION BY document_id, DATE(submission_timestamp) ORDER BY submission_timestamp) = 1
+),
+stable AS (
+  SELECT
+    DATE(submission_timestamp) AS submission_date,
+    document_id,
+  FROM
+    `moz-fx-data-shared-prod.firefox_installer_stable.install_v1`
+  WHERE
+    DATE(submission_timestamp) BETWEEN "2024-12-30" AND "2025-01-22"
+)
+
+SELECT
+  backfill.*,
+FROM
+  backfill
+LEFT JOIN
+  stable
+ON
+  backfill.document_id = stable.document_id
+  AND DATE(backfill.submission_timestamp) = stable.submission_date
+WHERE
+  stable.submission_date IS NULL
+```
+This inserts 12 171 249 rows into `moz-fx-data-backfill-1.firefox_installer_stable.install_v1`.
+
+## Insert into prod stable table
+
+This step requires DSRE assistance to write to the stable table (TODO: create a ticket).  The backfill staging table
+is deduped, so to insert into the prod table, we can run a `SELECT *` with a deestination table:
+```
+bq query \
+  --nouse_legacy_sql \
+  --destination_table 'moz-fx-data-shared-prod:firefox_installer_stable.install_v1' \
+  --append_table=true \
+  'SELECT * FROM `moz-fx-data-backfill-1.firefox_installer_stable.install_v1` WHERE DATE(submission_timestamp) BETWEEN "2024-12-30" AND "2025-01-22"'
+```
+
+Alternatively, we could use an `INSERT` statement, but that requires the columns in both tables to be in the same order,
```

Comment:
Ok, this is much better than generating the `INSERT` statement with the columns in the right order! Ended up doing that in https://github.com/mozilla/bigquery-backfill/blob/main/backfill/2024-11-06-metrics-pings-metric-labels-over-max-length/insert.sql